codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2ConnectionEncoder.java

```diff
@@ -416,13 +416,13 @@ public void write(ChannelHandlerContext ctx, int allowedBytes) {
                         //
                         // See https://github.com/netty/netty/issues/8707.
                         padding = dataSize = 0;
-                        return;
+                    } else {
+                        // There's no need to write any data frames because there are only empty data frames in the
+                        // queue and it is not end of stream yet. Just complete their promises by getting the buffer
+                        // corresponding to 0 bytes and writing it to the channel (to preserve notification order).
+                        ChannelPromise writePromise = ctx.newPromise().addListener(this);
```

Comment:
Wonder if we want to move the `addListener()` after the `write(...)` as well. Right now it's skewing the stack should the `remove(...)` complete the promise.

Reply:
Not sure I understand... can you show me some code ?

Reply:
Something like that...

```java
ChannelPromise writePromise = ctx.newPromise();
ctx.write(queue.remove(0, writePromise), writePromise)
    .addListener(this);
```

Otherwise if `queue.remove(0, writePromise)` was to complete the promise it'd potentially notify the listener first, followed by calling `ctx.write(...)`. Stack (traces) will look strange. Not sure there would be other side effects due to the order in which things execute but I think you want the listener to get notified after the write.

Reply:
Got it... I would like to investigate as a followup here. Trying to keep changes as minimal as possible atm